### PR TITLE
[gnmi] Surface py_gnmicli failures instead of hiding them as empty assertions (#26143)

### LINKED ISSUE #26143: [Test gap] gnmi helper.py swallows py_gnmicli failures — empty assertion body hides the real error
## Summary

Several helpers in `tests/gnmi/helper.py` run `gnmi_cli_py/py_gnmicli.py` on the PTF with `module_ignore_errors=True` and then read only `output['stdout']` — they never check `rc` / `failed` / `stderr`. When the client fails (exits non-zero with empty stdout), the failure is reported as an **empty assertion body** with no root cause, instead of the client's actual traceback. This turns any client-side crash into a misleading server/DB-looking failure and costs triage time.

This is a **test-logging gap**, independent of *why* the client failed. It should be fixed regardless of the underlying transport/environment bug.

- **Is it platform specific:** generic
- **Importance / Severity:** Medium (test reliability / triage-ability)
- **Branches seen:** master / 202511 / 202605 (the helper code is the same on all)

## Affected helpers

All read only `stdout` and never check `rc` / `stderr` (`tests/gnmi/helper.py` @ master):

- [`gnmi_get`](https://github.com/sonic-net/sonic-mgmt/blob/master/tests/gnmi/helper.py#L290-L334) — swallow point at [L334](https://github.com/sonic-net/sonic-mgmt/blob/master/tests/gnmi/helper.py#L334) (`raise Exception("error:" + msg)` with `msg == ''`).
- [`gnmi_subscribe_streaming_sample`](https://github.com/sonic-net/sonic-mgmt/blob/master/tests/gnmi/helper.py#L428) (`return msg, output['stderr']`, no check).
- [`gnmi_subscribe_streaming_onchange`](https://github.com/sonic-net/sonic-mgmt/blob/master/tests/gnmi/helper.py#L467) (same).
- [`gnmi_subscribe_polling`](https://github.com/sonic-net/sonic-mgmt/blob/master/tests/gnmi/helper.py#L384) (same gap; note it runs the Go `gnmi_cli` binary on the DUT, so it is not hit by the py_gnmicli import crash, but still returns raw output unchecked).

[`gnmi_set`](https://github.com/sonic-net/sonic-mgmt/blob/master/tests/gnmi/helper.py#L277-L287) and `test_gnmi_latency_01` already check the return code and surface the error correctly — they are the pattern to copy.

## Symptom

With the underlying client crash hidden, the nightly summary shows one of:

```
Exception: error:
```
```
AssertionError: Expected count of '<field>' not met. Expected count: N Actual count: 0 Message:
```

The defining trait is that **nothing follows** `error:` / `Message:` — no response text, no traceback. The real reason only appears with `--showlocals` or in the ansible DEBUG log, where the discarded `output` dict is visible — e.g. the client crashing at import:

```
msg    = ''
output = {'failed': True, 'rc': 1, 'stdout': '',
          'stderr': 'Traceback (most recent call last):
                       File ".../gnxi/gnmi_cli_py/py_gnmicli.py", line 46, in <module>
                         import gnmi_pb2
                       ...
                       ...gencode 7.35.0 runtime 6.33.5. Runtime version cannot be
                       older than the linked gencode version. See Protobuf version
                       guarantees at https://protobuf.dev/support/cross-version-runtime-guarantee.'}
```

## Root cause

In `tests/gnmi/helper.py`:

- `gnmi_get` reads only `output['stdout']`; empty stdout → no response marker → `raise Exception("error:" + msg)` with `msg=''`.
- the subscribe helpers `return (stdout, stderr)` with no failure check, and callers do `msg, _ = ...` — discarding stderr — then assert on `msg.count(<field>)`, which is `0` against an empty string.
- `ptfhost.shell(cmd, module_ignore_errors=True)` suppresses the ansible non-zero-rc failure, so the crash never surfaces as a shell error.

So the failing count is not a real "short response" — the client produced nothing, and the reason (in `stderr` + `rc`) is thrown away. An uncaught exception in `py_gnmicli.py` makes the Python interpreter write the traceback to **stderr** and exit **non-zero**, while the helper only inspects **stdout** (which stays empty because the crash happens before any response is printed).

## Steps to reproduce

Run a gnmi test that uses `gnmi_get` / `gnmi_subscribe_streaming_sample` (e.g. `gnmi/test_gnmi_countersdb.py::test_gnmi_queue_buffer_cnt` or `test_gnmi_counterdb_streaming_sample_01`) on a PTF where `py_gnmicli.py` exits non-zero with empty stdout (any import crash, timeout, or handshake failure that reports on stderr). Observed on 202511 and 202605 nightly PTF images. The test fails with an empty `Exception: error:` / `Actual count: 0 Message:` and no root cause in the summary.

## Suggested fix

Apply the same `rc`/`stderr`-aware check that [`gnmi_set`](https://github.com/sonic-net/sonic-mgmt/blob/master/tests/gnmi/helper.py#L277-L287) already uses — read `stderr`, gate on `rc != 0`, and put both in the raised exception — instead of folding empty stdout into a count assertion:

```python
output = ptfhost.shell(cmd, module_ignore_errors=True)
stdout = output.get("stdout") or ""
stderr = output.get("stderr") or ""
rc = output.get("rc", 1)
combined = f"{stdout}\n{stderr}"
if rc != 0 or "GRPC error" in combined or "rpc error" in combined:
    dump_gnmi_log(duthost)
    dump_system_status(duthost)
    raise Exception("py_gnmicli failed rc={}\nSTDOUT:\n{}\nSTDERR:\n{}".format(rc, stdout, stderr))
```

- **`gnmi_get`:** drop-in — it replaces the current empty `raise Exception("error:" + msg)` and keeps the existing GetResponse parsing.
- **subscribe helpers:** add the same guard before `return msg, output['stderr']`, and stop dropping stderr in callers (use `msg, err = ...` and include `err` in the assert message).

**Caveat for the subscribe helpers:** unlike `gnmi_set` (which returns nothing), callers of the subscribe helpers assert on `msg.count(<field>)`, so a legitimately-empty-but-successful response (`rc == 0`, empty stderr) must still flow back to the caller. Only raise on the hard-failure signal (`rc != 0` or a `Traceback` / `GRPC error` / `rpc error` in `combined`) — that catches the import crash and connection/handshake failures while leaving valid empty results for the caller's own assertion.

## Related

- **Current trigger:** a PTF protobuf gencode/runtime mismatch makes `py_gnmicli.py` crash at import (`import gnmi_pb2`, `gencode 7.35.0` vs `runtime 6.33.5`). That environment bug is what this logging gap is currently hiding; this issue is specifically about the test helper not surfacing such failures.
- **Prior art (same crash class, already fixed once):** #23273 — "PTF protobuf version mismatch" (202505, closed). It fixed the version in sonic-buildimage but did not address this helper logging gap.

### PR DESCRIPTION
**Pull request by yxieca**
### Description of PR

Summary:
Fixes #26143

Several helpers in `tests/gnmi/helper.py` run `gnmi_cli_py/py_gnmicli.py` with
`module_ignore_errors=True` and then inspect only `output['stdout']` — they never
check `rc` / `stderr`. When the client exits non-zero with empty stdout (e.g. an
import crash, handshake failure, or protobuf gencode/runtime mismatch), the real
traceback lives in `stderr`/`rc` and gets discarded. The failure then surfaces as
a misleading empty assertion body:

```
Exception: error:
AssertionError: Expected count of '<field>' not met. Expected count: N Actual count: 0 Message:
```

with no root cause, costing triage time and making a client-side crash look like a
server/DB failure. This is a generic test-logging gap; the helper code is identical
on master / 202511 / 202605.

This PR applies the same `rc`/`stderr`-aware check that `gnmi_set` already uses to
the affected helpers, via a shared `_raise_on_gnmi_cli_failure()` helper:

- `gnmi_get` — now surfaces `rc`/`stderr` on failure instead of raising an empty
  `error:` when stdout has no `GetResponse` marker.
- `gnmi_subscribe_streaming_sample`
- `gnmi_subscribe_streaming_onchange`
- `gnmi_subscribe_polling` (runs the Go `gnmi_cli` on the DUT; guarded for symmetry)

Only hard-failure signals raise (`rc != 0`, or a `Traceback` / `GRPC error` /
`rpc error` in the combined output), so a legitimately-empty-but-successful
response (`rc == 0`, empty stderr) still flows back to the caller for its own
`msg.count(<field>)` assertion — this preserves existing subscribe-helper behavior
while catching import crashes and connection/handshake failures.

### Type of change

- [x] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] New Test case
- [ ] Test case improvement

#### What is the motivation for this PR?

Client-side `py_gnmicli.py` crashes were being hidden behind empty assertion
messages, misdirecting triage toward the server/DB. This makes gnmi test failures
self-explaining by including the client's actual `rc` and `stderr` in the raised
exception, without changing any pass/fail semantics for healthy runs.

#### How did you do it?

Added a shared `_raise_on_gnmi_cli_failure(duthost, output)` guard that reads
`stdout`/`stderr`/`rc`, dumps gnmi + system diagnostics, and raises with the full
context on hard failure. Wired it into the four helpers that previously read only
stdout. Left the valid-empty-response path intact for subscribe callers.

#### How did you verify/test it?

- `python3 -m py_compile tests/gnmi/helper.py` — OK
- `flake8 --max-line-length=120 tests/gnmi/helper.py` — clean
- Logic-reviewed each caller to confirm valid empty (rc==0) responses still return
  to the caller and only hard failures raise.

### Documentation

No.

### CODE REVIEW

**Comment by yxieca** [comment]
This PR addresses test-gap issue #26143. It applies the `rc`/`stderr`-aware failure check that `gnmi_set` already uses to `gnmi_get`, `gnmi_subscribe_streaming_sample`, `gnmi_subscribe_streaming_onchange`, and `gnmi_subscribe_polling`, so a `py_gnmicli.py` crash surfaces its real traceback instead of an empty `error:` / `Actual count: 0` assertion. Valid empty-but-successful responses (`rc == 0`, empty stderr) still flow back to callers unchanged.

🤖 *This response was generated by an AI agent on behalf of @yxieca.*

**Comment by yxieca** [comment]
@ronan-nexthop @liang-nexthop please help review. Thanks.

**Review by ronan-nexthop — APPROVED** [review]
LGTM, this infra will be replaced soon with better testing infra using `pygnmi` which will fix all this, this is good for the interim

File: tests/gnmi/helper.py [hunk 1]
```diff
@@ -424,6 +462,7 @@ def gnmi_subscribe_streaming_sample(duthost, ptfhost, path_list, interval_ms, co
         path = path.replace('sonic-db:', '')
         cmd += " " + path
     output = ptfhost.shell(cmd, module_ignore_errors=True)
+    _raise_on_gnmi_cli_failure(duthost, output)
```
Comment (hdwhdw) [inline tests/gnmi/helper.py]: [P1] This guard turns an expected authorization denial into an exception before its caller can validate it. `test_gnmi_configdb_subscribe_authenticate` assigns `gnmi_config_db_noaccess`, then expects this helper to return output containing `GRPC error` and the role; `_raise_on_gnmi_cli_failure()` now raises on that exact response. Distinguish transport or client failures from valid server RPC statuses and preserve denial output for callers. Add a no-access subscription test that returns the denial for assertion, plus a client-crash test that still raises with `rc` and stderr.

**Review by hdwhdw — COMMENTED** [review]
Centralizing CLI failure diagnostics is useful, but this guard changes an existing caller contract. It treats a valid authorization denial as a client crash and breaks the no-access subscription test. The head conflicts with current master; static checks passed, but no clean current-base runtime result exists.

File: tests/gnmi/helper.py (hunk 1 above)
Comment (hdwhdw) [inline tests/gnmi/helper.py:465]: **❌ [P1] Preserve expected denial responses**

This guard breaks an existing authorization route. T0 and T1-LAG failed `test_gnmi_configdb_subscribe_authenticate[vrf_default_1]` three times. `py_gnmicli` returned `rc=1` with the expected `gnmi_config_db_noaccess` response, then this line raised before the caller could validate it.

Keep non-empty protocol responses available to callers. Limit the shared guard to client failures that would otherwise be hidden. For example:

```python
stdout = output.get("stdout") or ""
stderr = output.get("stderr") or ""
rc = output.get("rc", 1)

client_crash = "Traceback (most recent call last)" in stderr
missing_response = rc != 0 and not stdout.strip()
if client_crash or missing_response:
    raise Exception(...)
```

Add a no-access test that returns the denial for caller validation. Add a client-crash test that still raises for empty output.

**Review by hdwhdw — COMMENTED** [review]
Almost LGTM—one caller-contract regression needs a fix before approval.

- Centralizing hidden client failures is the right direction.
- T0 and T1-LAG reproduced the denial-path failure three times.
- T2 and multi-ASIC T1 skipped this case.
- The head conflicts with current `master`; rerun after resolving it.
